app/src/main/java/com/cmput301f20t21/bookfriends/controller/SignUpController.java

```diff
@@ -0,0 +1,66 @@
+package com.cmput301f20t21.bookfriends.controller;
+
+import androidx.annotation.NonNull;
+
+import com.cmput301f20t21.bookfriends.services.AuthService;
+import com.cmput301f20t21.bookfriends.services.UserService;
+import com.google.android.gms.tasks.OnCompleteListener;
+import com.google.android.gms.tasks.Task;
+import com.google.firebase.auth.AuthResult;
+import com.google.firebase.firestore.QuerySnapshot;
+
+public class SignUpController {
+    private AuthService authService;
+    private UserService userService;
+    private ISignUpListener listener;
+
+    public interface ISignUpListener {
+        void onSignUpSuccess();
+        void onSignUpFail();
+    }
+
+    public SignUpController(ISignUpListener listener) {
+        authService = AuthService.getInstance();
+        userService = UserService.getInstance();
+        this.listener = listener;
+    }
+
+    public void handleSignUp(final String username, final String email, final String password) {
+        // check if the email has been registered by another user
+        userService.getUserByEmail(email).addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
```

Comment:
do you guys think we should check this?
current flow:
check email -> check username exists -> create username -> create auth
Or do we want:
check username exists -> create auth -> if success -> create username with email

Reply:
the second option seems less work to me, since create auth will reject if the email already exist?

Reply:
Yeah, I also prefer second option.

Reply:
will update in view model refactor PR.